### PR TITLE
[FEATURE] Add TLS support.

### LINKED ISSUE #368: Authelia support TLS
It would be nice to have Authelia support TLS to raise the level of security between the proxy and Authelia.

### PR DESCRIPTION
**Pull request by clems4ever**
Fixes #368.

### CODE REVIEW

File: docs/configuration/miscellaneous.md [hunk 1]
```diff
@@ -10,13 +10,25 @@ nav_order: 5
...
 
     host: 0.0.0.0
     port: 9091
 
+## TLS
+
+`optional: true`
+
+Authelia can use TLS. Provide the certificate and the key with the
+following keys.
```
Comment (nightah) [inline docs/configuration/miscellaneous.md]:
```suggestion
following configuration keys.
```

File: internal/configuration/validator/configuration.go [hunk 2]
```diff
@@ -24,6 +24,12 @@ func Validate(configuration *schema.Configuration, validator *schema.StructValid
 		configuration.LogsLevel = defaultLogsLevel
 	}
 
+	if configuration.SSLKey != "" && configuration.SSLCert == "" {
+		validator.Push(fmt.Errorf("SSL key provided but no certificate"))
```
Comment (nightah) [inline internal/configuration/validator/configuration.go]: I feel like we could be much more concise in our messaging here. It doesn't matter if they have done something correct we should explicitly be calling out what wasn't done and potentially is causing issues.

```suggestion
		validator.Push(fmt.Errorf("No SSL certificate provided, please check the "ssl_cert" which has been configured"))
```

File: internal/configuration/validator/configuration.go [hunk 3]
```diff
@@ -24,6 +24,12 @@ func Validate(configuration *schema.Configuration, validator *schema.StructValid
...
+	} else if configuration.SSLKey == "" && configuration.SSLCert != "" {
+		validator.Push(fmt.Errorf("SSL certificate provided but no key"))
```
Comment (nightah) [inline internal/configuration/validator/configuration.go]: Same as above.

```suggestion
		validator.Push(fmt.Errorf("No SSL keyprovided, please check the "ssl_key" which has been configured"))
```

File: internal/configuration/validator/configuration_test.go [hunk 4]
```diff
@@ -99,3 +99,33 @@ func TestShouldAddDefaultAccessControl(t *testing.T) {
...
 }
+
+func TestShouldRaiseErrorWhenSSLCertWithoutKeyIsProvided(t *testing.T) {
+	validator := schema.NewStructValidator()
+	config := newDefaultConfig()
+	config.SSLCert = "/tmp/cert.pem"
+
+	Validate(&config, validator)
+	require.Len(t, validator.Errors(), 1)
+	assert.EqualError(t, validator.Errors()[0], "SSL certificate provided but no key")
```
Comment (nightah) [inline internal/configuration/validator/configuration_test.go]: Update per above recommendations:

```suggestion
	assert.EqualError(t, validator.Errors()[0], "No SSL key provided, please check the "ssl_key" which has been configured")
```

File: internal/configuration/validator/configuration_test.go [hunk 5]
```diff
@@ -99,3 +99,33 @@ func TestShouldAddDefaultAccessControl(t *testing.T) {
...
+}
+
+func TestShouldRaiseErrorWhenSSLKeyWithoutCertIsProvided(t *testing.T) {
+	validator := schema.NewStructValidator()
+	config := newDefaultConfig()
+	config.SSLKey = "/tmp/key.pem"
+
+	Validate(&config, validator)
+	require.Len(t, validator.Errors(), 1)
+	assert.EqualError(t, validator.Errors()[0], "SSL key provided but no certificate")
```
Comment (nightah) [inline internal/configuration/validator/configuration_test.go]:
```suggestion
	assert.EqualError(t, validator.Errors()[0], "No SSL certificate provided, please check the "ssl_cert" which has been configured")
```

File: internal/suites/Standalone/users.yml [hunk 6]
```diff
@@ -13,7 +13,7 @@ users:
     email: james.dean@authelia.com
     groups: []
   john:
-    password: $6$rounds=500000$jgiCMRyGXzoqpxS3$w2pJeZnnH8bwW3zzvoMWtTRfQYsHbWbD/hquuQ5vUeIyl9gdwBIt6RWk2S6afBA0DPakbeWgD/4SZPiS0hYtU/
+    password: '{CRYPT}$6$rounds=50000$LnfgDsc2WD8F2qNf$0gcCt8jlqAGZRv2ee3mCFsfAr1P4N7kESWEf36Xtw6OjkhAcQuGVOBHXp0lFuZbppa7YlgHk3VD28aSQu9U9S1'
```
Comment (nightah) [inline internal/suites/Standalone/users.yml]: I assume you added this back in to ensure that both variants with and without the `{CRYPT]` worked?

Comment (clems4ever) [reply internal/suites/Standalone/users.yml]: damn it has been updated by the suite I guess, this change was not intended.

File: internal/suites/Standalone/configuration.yml [hunk 7]
```diff
@@ -3,6 +3,8 @@
 ###############################################################
 
 port: 9091
+ssl_cert: /var/lib/authelia/ssl/cert.pem
```
Comment (clems4ever) [inline internal/suites/Standalone/configuration.yml]: After second thought, I'll replace `ssl_cert` and `ssl_key` by `tls_cert` and `tls_key`.

**Comment by clems4ever** [comment]
@nightah, your latest comment relates to a previous commit. Otherwise, all changes have been done on my side.

**Pull request merged**